(zstd compression?) com.google.common.base.VerifyException: java.io.IOException: attempt to write to a closed Outputstream backed by a native file

### Description of the bug:


We're observing sporadic build failures where the Bazel daemon crashes with the following:

```
240630 09:50:38.941:XT 2149 [com.google.devtools.build.lib.bugreport.BugReport.handleCrash] Handling crash with CrashContext{haltJvm=true, args=[], sendBugReport=true, extraOomInfo=, eventHandler=com.google.devtools.build.lib.events.Reporter@2a243856}
com.google.common.base.VerifyException: java.io.IOException: attempt to write to a closed Outputstream backed by a native file
        at com.google.devtools.build.lib.remote.GrpcCacheClient$1.onNext(GrpcCacheClient.java:433)
        at com.google.devtools.build.lib.remote.GrpcCacheClient$1.onNext(GrpcCacheClient.java:414)
        at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onMessage(ClientCalls.java:474)
        at io.grpc.ForwardingClientCallListener.onMessage(ForwardingClientCallListener.java:33)
        at io.grpc.ForwardingClientCallListener.onMessage(ForwardingClientCallListener.java:33)
        at com.google.devtools.build.lib.remote.logging.LoggingInterceptor$LoggingForwardingCall$1.onMessage(LoggingInterceptor.java:138)
        at io.grpc.internal.DelayedClientCall$DelayedListener$2.run(DelayedClientCall.java:457)
        at io.grpc.internal.DelayedClientCall$DelayedListener.drainPendingCallbacks(DelayedClientCall.java:507)
        at io.grpc.internal.DelayedClientCall$1DrainListenerRunnable.runInContext(DelayedClientCall.java:296)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.io.IOException: attempt to write to a closed Outputstream backed by a native file
        at com.google.devtools.build.lib.unix.UnixFileSystem$NativeFileOutputStream.write(UnixFileSystem.java:562)
        at com.google.devtools.build.lib.remote.common.LazyFileOutputStream.write(LazyFileOutputStream.java:44)
        at com.google.devtools.build.lib.remote.RemoteCache$ReportingOutputStream.write(RemoteCache.java:550)
        at com.google.devtools.build.lib.remote.util.DigestOutputStream.write(DigestOutputStream.java:58)
        at com.google.common.io.CountingOutputStream.write(CountingOutputStream.java:54)
        at com.google.devtools.build.lib.remote.zstd.ZstdDecompressingOutputStream.write(ZstdDecompressingOutputStream.java:61)
        at com.google.devtools.build.lib.remote.zstd.ZstdDecompressingOutputStream.write(ZstdDecompressingOutputStream.java:54)
        at com.google.protobuf.ByteString$LiteralByteString.writeTo(ByteString.java:1459)
        at com.google.devtools.build.lib.remote.GrpcCacheClient$1.onNext(GrpcCacheClient.java:430)
        ... 12 more
```

### Which category does this issue belong to?

Remote Execution

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.


_No response_

### Which operating system are you running Bazel on?


Linux

### What is the output of `bazel info release`?


release 7.2.0-vmware

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.


_No response_

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?


_No response_

### If this is a regression, please try to identify the Bazel commit where the bug was introduced with bazelisk --bisect.


_No response_

### Have you found anything relevant by searching the web?

No. I couldn't find any matches for "closed outputstream".

### Any other information, logs, or outputs that you want to share?

We're using dynamic RBE (impl: bazelbuild/bazel-buildfarm) and remote cache compression.

```
build:remote --extra_execution_platforms=//:rbe_platform
build:remote --remote_executor=grpcs://<endpoint>
build:remote --jobs=HOST_CPUS*10
build:remote --remote_retries=5
build:remote --experimental_remote_cache_eviction_retries=5
build:remote --verbose_failures
build:remote --remote_cache=
build:remote --disk_cache=
build:remote --noremote_upload_local_results
build:remote --experimental_remote_cache_async
build:remote --experimental_remote_merkle_tree_cache
build:remote --remote_local_fallback
build:remote --remote_local_fallback_strategy=sandboxed
build:remote --experimental_remote_downloader_local_fallback
build:remote --remote_cache_compression

build:rbe_dynamic \
    --config=remote \
    --internal_spawn_scheduler \
    --spawn_strategy=dynamic \
    --dynamic_local_strategy=worker,sandboxed,local

build --experimental_debug_spawn_scheduler
```

--- a/src/main/java/com/google/devtools/build/lib/remote/GrpcCacheClient.java
+++ b/src/main/java/com/google/devtools/build/lib/remote/GrpcCacheClient.java
@@ -42,7 +42,6 @@ import com.google.bytestream.ByteStreamProto.ReadResponse;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Ascii;
 import com.google.common.base.Preconditions;
-import com.google.common.base.VerifyException;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Iterables;
 import com.google.common.flogger.GoogleLogger;
@@ -483,8 +482,11 @@ public class GrpcCacheClient implements RemoteCacheClient, MissingDigestsFinder 
                 .setReadOffset(rawOut.getCount())
                 .build(),
             new ClientResponseObserver<ReadRequest, ReadResponse>() {
+              private volatile ClientCallStreamObserver<ReadRequest> requestStream;
+
               @Override
               public void beforeStart(ClientCallStreamObserver<ReadRequest> requestStream) {
+                this.requestStream = requestStream;
                 future.addListener(
                     () -> {
                       if (future.isCancelled()) {
@@ -500,8 +502,13 @@ public class GrpcCacheClient implements RemoteCacheClient, MissingDigestsFinder 
                 try {
                   data.writeTo(out);
                 } catch (IOException e) {
-                  // Cancel the call.
-                  throw new VerifyException(e);
+                  // The output stream was likely closed due to cancellation (e.g. dynamic execution
+                  // choosing the local branch).
+                  if (requestStream != null) {
+                    requestStream.cancel("output stream closed", e);
+                  }
+                  future.setException(e);
+                  return;
                 }
                 // reset the stall backoff because we've made progress or been kept alive
                 progressiveBackoff.reset();

